File: qa/rgw/s3-tests/s3tests/functional/policy.py
import json

class Statement(object):
    def __init__(self, action, resource, principal = {"AWS" : "*"}, effect= "Allow", condition = None):
        self.principal = principal
        self.action = action
        self.resource = resource
        self.condition = condition
        self.effect = effect

    def to_dict(self):
        d = { "Action" : self.action,
              "Principal" : self.principal,
              "Effect" : self.effect,
              "Resource" : self.resource
        }

        if self.condition is not None:
            d["Condition"] = self.condition

        return d

class Policy(object):
    def __init__(self):
        self.statements = []

    def add_statement(self, s):
        self.statements.append(s)
        return self

    def to_json(self):
        policy_dict = {
            "Version" : "2012-10-17",
            "Statement":
            [s.to_dict() for s in self.statements]
        }

        return json.dumps(policy_dict)

def make_json_policy(action, resource, principal={"AWS": "*"}, conditions=None):
    """
    Helper function to make single statement policies
    """
    s = Statement(action, resource, principal, condition=conditions)
    p = Policy()
    return p.add_statement(s).to_json()

File: src/test/rgw/s3-tests-rs/src/policy.rs
use serde_json::{json, Value};

pub fn make_json_policy(
    action: &str,
    resource: &str,
    principal: Option<Value>,
    effect: Option<&str>,
    conditions: Option<Value>,
) -> String {
    let principal = principal.unwrap_or(json!({"AWS": "*"}));
    let effect = effect.unwrap_or("Allow");

    let mut statement = json!({
        "Action": action,
        "Principal": principal,
        "Effect": effect,
        "Resource": resource,
    });

    if let Some(cond) = conditions {
        statement["Condition"] = cond;
    }

    let policy = json!({
        "Version": "2012-10-17",
        "Statement": [statement],
    });

    serde_json::to_string(&policy).unwrap()
}

pub fn make_arn_resource(path: &str) -> String {
    format!("arn:aws:s3:::{path}")
}
